### PR TITLE
Add test for step data update with CLI options

### LINKED ISSUE #2016: `tmt plan show` doesn't list provision details for /plans/provision/beaker
The idea is that without verbosity options only non-default values are printed. However provision is set to 'how: beaker' in this plan but isn't printed. 

@psss found out that this behaviour change happened with 68bee3e

To reproduce: `tmt plans show /plans/provision/beaker` and see that `provision` is not listed in the output. In the `-v` mode one sees that non-default value of `how: beaker` should be used (the default provision method is how: virtual)

### PR DESCRIPTION
**Pull request by happz**
This functionality has been inside `BasePlugin`, and today a bug was reported with its handling of flag-like options. Therefore, this patch:

* extracts the code into a standalone function, complementing the rest of normalization code,
* fixes the bug,
* fixes a few more issues discovered, mostly readability,
* adds a pile of comment WRT various related issues, and
* adds a unit test.

### CODE REVIEW

File: tmt/utils.py [hunk 1]
```diff
@@ -3857,6 +3857,75 @@ def dataclass_normalize_field(
...
+    """
+    Normalize and assign values from CLI options to container fields.
+
+    :param container: container to update.
+    :param option_getter: a callable to provide a value for a given option.
+        It is called once for each option name, with an optional default
+        to return if the option was not specified.
+    :param keys: optional list of keys to update. If not set, all keys
+        are considered.
+    :param preserve_modified: if set, only options whose value is not equal
```
Comment (thrix) [inline tmt/utils.py:3884]: logger is missing from params

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/utils.py [hunk 2]
```diff
@@ -3857,6 +3857,75 @@ def dataclass_normalize_field(
...
+        to field's default value would be saved in the container.
+    """
+
+    keys = keys or list(container.keys())
+
+    for keyname in keys:
+        # Checks below try to prevent default options of CLI options from
+        # overwriting values already consumed from fmf nodes.This is no
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/utils.py]: Fixed in 00107c3e

File: tmt/utils.py (hunk 1 above)
Comment (happz) [reply tmt/utils.py:3886]: Fixed in 00107c3e

**Review by thrix — APPROVED** [review]
LGTM, test failed on some network glitch, restarted

**Review by psss — CHANGES_REQUESTED** [review]
The default handling does not seem to be ok.

File: tmt/utils.py [hunk 3]
```diff
@@ -3859,6 +3859,76 @@ def dataclass_normalize_field(
...
+        if value is dataclasses.MISSING:
+            continue
+
+        if preserve_modified:
+            # If the value we get for an option is the default value, good,
+            # no need to bother - as such, it's either already stored in
+            # `self.data`, or it has been replaced with whatever came from an
+            # fmf node.
+            if value == container._default(keyname):
+                continue
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/utils.py:3915]: All of this is just wrong... It's all guessing, and it's a big smoking pile of corner cases. Let me find out what exactly breaks when this particular check is dropped, I bet there's something just waiting to bite us :)

File: tests/unit/test_dataclasses.py [hunk 4]
```diff
@@ -152,3 +153,58 @@ class DummyContainer(SerializableContainer):
...
+
+
+def test_normalize_options(root_logger):
+    @dataclasses.dataclass
+    class DummyContainer(DataContainer):
+        foo: int = field(option='--foo', default=1)
+        bar: List[str] = field(option='--bar', default_factory=list, multiple=True)
+        baz: str = 'undefined'
+        qux: List[str] = field(option='--qux', default_factory=list, multiple=True)
+        corge: bool = field(option='--corge / --no-corge', default=True, multiple=True)
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/unit/test_dataclasses.py:165]: Hmm, depends, could be. `--foo --no-foo --foo` might look pointless, but maybe there's a use for it. It is definitely a copy & paste issue rather than intention.

File: tmt/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (psss) [reply tmt/utils.py:3915]: Summary from the call: We need to find a reliable way how to detect if a flag option was provided on the command line or not (versus set as the default). Moving to the next release because of the regression (unable to override non-default config value with default value provided on the command line).

[the remark this answers is not in the document]
Comment (happz) [reply tmt/utils.py:3915]: Right, it's the boolean value of these defaults. TL;DR, there's no way for code to tell whether `-i` was used on the command line, or whether the `fedora` string comes from the default value. For the code, these two look the same: `-h container` vs `-h container -i fedora`, both would present `fedora` as the value of the `--image` CLI option. Now, in the first case, we do not want to overwrite fmf source, in the second we do want to. How to tell the difference?

As briefly discussed, there is a couple of options, none trivial or immediately available. I'd try to look into introducing a callback that would notify us when an option was actually encountered on a command line, and we'd then simply use the value no matter what it ould be. Postponing this PR to avoid bigger harm.

**Comment by psss** [comment]
Any plans with this one, @happz?

**Comment by psss** [comment]
Let's try to finish this one?

**Comment by therazix** [comment]
@happz Is this PR still relevant?
